model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/JpaMapKeycloakTransaction.java

```diff
@@ -348,28 +345,19 @@ private Predicate notExpired(final CriteriaBuilder cb, final JpaSubqueryProvider
 
     private static class QueryCacheKey {
         private final String queryString;
-        private final Integer queryMaxResults;
-        private final Integer queryFirstResult;
-        private final HashMap<String, Object> queryParameters;
+        private final Integer queryLimit;
+        private final Integer queryOffset;
         private final Class<?> modelType;
+        private final List<? extends QueryParameters.OrderBy<?>> queryOrderBy;
 
-        public QueryCacheKey(TypedQuery<?> emQuery, Class<?> modelType) {
+        public QueryCacheKey(QueryParameters<?> query, Class<?> modelType) {
             // copy over all fields from the query that relevant for caching
-            QueryImplementor<?> query = emQuery.unwrap(QueryImplementor.class);
-            this.queryString = query.getQueryString();
-            this.queryParameters = new HashMap<>();
-            for (Parameter<?> parameter : query.getParameters()) {
-                // HACK TO GET KC TO START WITH NEW STORE - HAS TO BE FIXED AS SOME SEARCHES USE A DIFFERENT EXPRESSIBLE IMPL
-                JpaCriteriaParameter criteriaParameter = (JpaCriteriaParameter) parameter;
-                SqmBasicValuedSimplePath expressible = (SqmBasicValuedSimplePath) criteriaParameter.getExpressible();
-                String parameterName  = expressible.getNavigablePath().getLocalName();
-                if (parameterName == null) {
-                    throw new ModelException("Can't prepare query for caching as parameter doesn't have a name");
-                }
-                this.queryParameters.put(parameterName, ((JpaCriteriaParameter<?>) parameter).getValue());
-            }
-            this.queryMaxResults = emQuery.getMaxResults();
-            this.queryFirstResult = emQuery.getFirstResult();
+            DefaultModelCriteria dmc = query.getModelCriteriaBuilder()
```

Comment:
Nit: this could be checked to be `DefaultModelCriteria` already, in which case flashing could be skipped.

Reply:
I agree